pkg/crawler/central/central.go

```diff
@@ -0,0 +1,376 @@
+
+	// Fetch SHA1s from Maven Central
+	if err := s.fetchSHA1s(ctx, recordCh); err != nil {
+		return xerrors.Errorf("failed to fetch SHA1s: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Source) read(ctx context.Context, records <-chan data.Record, errCh chan error, recordCh chan<- types.Record) error {
+	var record data.Record
+	var ok bool
+	var count int
+	for {
+		select {
+		case record, ok = <-records:
+			if !ok {
+				return nil
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-errCh:
+			return err
+		}
+
+		if record.Type() != data.ArtifactAdd {
+			continue
+		}
+
+		// Since SHA1 is not reliable in central index, we need to fetch it from Maven Central
+		// cf. https://github.com/aquasecurity/trivy-java-db/pull/58#issuecomment-2890441162
+		rec := types.Record{
+			GroupID:    mustGet[string](record, "groupId"),
+			ArtifactID: mustGet[string](record, "artifactId"),
+			Version:    mustGet[string](record, "version"),
+			Classifier: mustGet[string](record, "classifier"),
+		}
+
+		switch {
+		case rec.GroupID == "" || rec.ArtifactID == "" || rec.Version == "":
+			continue
+		case containsControlChar(rec.GroupID + rec.ArtifactID + rec.Version + rec.Classifier):
+			// Skip records with control characters in groupId, artifactId, version or classifier
+			continue
+		case !maven.ValidateClassifier(rec.Classifier):
+			continue
+		}
+
+		// Skip if already processed
+		gavcHash := hash.GAVC(rec.GroupID, rec.ArtifactID, rec.Version, rec.Classifier)
+		if _, exists := s.storedGAVCs[gavcHash]; exists {
+			continue
+		}
+		s.storedGAVCs[gavcHash] = struct{}{}
+		s.records = append(s.records, rec)
+
+		count++
+		if count%100000 == 0 {
+			s.logger.Info(fmt.Sprintf("Parsed %d records", count))
+		}
+	}
+}
+
+// fetchSHA1s fetches SHA1s from Maven Central repository
+//
+// For each GAVC record collected from the central index, this function constructs the corresponding
+// Maven Central URL and fetches the SHA1 checksum directly. This avoids using the unreliable SHA1s
+// from the index itself.
+func (s *Source) fetchSHA1s(ctx context.Context, recordCh chan<- types.Record) error {
+	if len(s.records) == 0 {
+		return nil
+	}
+
+	// Create an error group for parallel fetching
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(100)
+
+	var processedCount, fetchedCount atomic.Int64
+
+	// Process each record
+	for _, rec := range s.records {
+		g.Go(func() error {
+			defer func() {
+				current := processedCount.Add(1)
+				if current%10000 == 0 {
+					s.logger.Info(fmt.Sprintf("Fetched %d SHA1s out of %d", current, len(s.records)))
+				}
+			}()
+
+			// Create the SHA1 URL based on GAV coordinates
+			// Maven Central URL pattern: baseURL/groupId/artifactId/version/artifactId-version[-classifier].jar.sha1
+			// Convert dots in groupId to slashes
+			groupPath := strings.ReplaceAll(rec.GroupID, ".", "/")
+
+			// Build jar name: artifactId-version[-classifier].jar
+			jarName := rec.ArtifactID + "-" + rec.Version + ".jar"
```

Comment:
classifiers are missed.

Reply:
There was a reason why I didn't add a classifier here, but I can't recall it...
Fixed in https://github.com/aquasecurity/trivy-java-db/pull/58/commits/69a21172a896a13dba3a50870b3b3631a542fcdb